src/ruby/puppetserver-lib/puppet/server/http_client.rb

```diff
@@ -96,7 +90,14 @@ def create_common_request_options(url, headers, options)
 
     # Ensure multiple requests are not made on the same connection
     headers["Connection"] = "close"
-    request_options = RequestOptions.new(build_url(url))
+
+    if url.is_a?(String)
+      url = URI(url)
```

Comment:
I believe the `url` needs to be URL encoded before being passed to the `URI(..)` parser. A good test is to pass in a URL with a space like `http://a.com/b c`. If the space is not encoded, then it will fail to parse:

```
irb(main):023:0> URI('http://a.com/b c')
Traceback (most recent call last):
        6: from /usr/local/opt/rbenv/versions/2.5.7/bin/irb:11:in `<main>'
        5: from (irb):23
        4: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.7/lib/ruby/2.5.0/uri/common.rb:743:in `URI'
        3: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.7/lib/ruby/2.5.0/uri/common.rb:237:in `parse'
        2: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.7/lib/ruby/2.5.0/uri/rfc3986_parser.rb:73:in `parse'
        1: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.7/lib/ruby/2.5.0/uri/rfc3986_parser.rb:67:in `split'
URI::InvalidURIError (bad URI(is not URI?): http://a.com/b c)
```

That said I'm curious why puppetserver would receive a String and not a URI? I wanted to define the API in terms of a URI object so that the implementation could be certain that encoding had already been performed (to avoid confusion like the string being double URL encoded)

Reply:
Most of our testing is passing strings right now. I'm not sure if there is anywhere in real code that passes strings, but I can look into it if we want to avoid that.

Reply:
@joshcooper so is it true that the puppet-agent implementation does not worth with a bare string, and I should explicitly disallow that here (that is, error if you pass one instead of coercing)? I can go through and update all of our tests (notably [these](https://github.com/puppetlabs/puppetserver/blob/6.x/test/unit/puppetlabs/puppetserver/ruby/http_client_test.clj)) to wrap their strings in `URI()` if that is the expectation.

Reply:
I did add a test just now that passes an unescaped string, and it fails as you say. But if we update this API to not accept strings, seems like we don't really need that test.

Reply:
> is it true that the puppet-agent implementation does not work with a bare string, and I should explicitly disallow that here

That's correct, it requires a URI object so that it's clear that URL encoding needs to be handled by the caller.